### ISSUE TITLE
AdSense not being connected after going through the initial setup

### ISSUE THREAD

**Issue by mohitwp** [google/site-kit-wp#5614]
## Bug Description

AdSense not being connected after going through the initial setup
This issue occurs once on fresh site. It occurs only one time while doing set up first time, after that this issue is not getting occur even after resetting the plugin.

Note : This issue is occurring when we try to do set up via clicking on Adsense CTA on dashboard or via settings page.

## Steps to reproduce

1. Create fresh site using dev plugin.
2. Do initial set up.
3. Try to set up AdSense through CTA on dashboard or via settings page.
4. Check Dashboard and settings page will show AdSense not connected.

## Screenshots

https://user-images.githubusercontent.com/1621608/179300930-e2337c4e-7270-406b-978e-a1081cafadc5.mp4

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Under successful preconditions for connecting AdSense, the module's setup should always complete successfully

## Implementation Brief

We need to add a flag to prevent the AdSense background `submitChanges()` action being triggered while the user is actively submitting the changes.

- Use this draft [PR](https://github.com/google/site-kit-wp/pull/7374) as a guide.
- However, use a less abstract name for the flag. Rather than `isCompletingSetup` etc., name the flag `suspendBackgroundSubmit` and its permutations.
- Review this [comment](https://github.com/google/site-kit-wp/issues/5614#issuecomment-1660123104) for some tips on reproducing the issue.

### Test Coverage

- No new tests needed.

## QA Brief

The AdSense module should not be set up/configured prior to the QA steps that follow. If it is, reset the Site Kit plugin and perform the QA steps that follow:

- Perform set up steps for the Google AdSense module via the `Google Site Kit > Settings` screen.
- Follow the UI prompts / CTA's to finalise set up of the AdSense service.
- Validate that the AdSense module is correctly set up and indicates as such in the `Google Site Kit > Settings` screen within the connected modules section.

Give that it is not straightforward to replicate this race condition in all environments, the steps above (using a reset of the plugin each time) should be performed ~10 times.

## Changelog entry

* Fix potential bug in AdSense set up which could leave the setup incomplete.

**Comment by wpdarren** [comment]
@mohitwp I've not been able to recreate this.  Did you have the  `adsenseSetupV2` feature flag enabled? Also, what other settings did you have enabled related to AdSense.  Was this discovered by while completing 1.80.0 release QA and could you recreate it on future versions of Site Kit? Sorry, a few questions but we could do re-testing this I feel.

**Comment by aaemnnosttv** [comment]
@wpdarren @mohitwp I have seen this before but was not able to reproduce consistently either. I have not seen it lately though so I'm wondering if it has been inadvertently fixed somewhere along the way 😄 

If you could both please give this a try again to reproduce, that would be great. If this is reproducible (even if not consistently) then we need to move this forward.

**Comment by mohitwp** [comment]
@aaemnnosttv @wpdarren  I'm able to reproduce this issue, but not consistently. It happens if after initial set up, user do not refresh dashboard or perform any other activity before setting up AdSense.

Verified using dev plugin and oi.ie site.
Feature flag enabled - adsensesetupV2, DashboardSharing, GA4 Activation banner.

- Create fresh site using dev plugin.
- Do initial set up.
- Do not refresh page or perform any activity like connect/disconnect module, sharing dashboard.
- Try to set up AdSense through CTA on dashboard or via settings page.
- Check Dashboard and settings page will show AdSense not connected.
- PHP 8.1

**Comment by wpdarren** [comment]
@mohitwp @aaemnnosttv I tried this on a few new sites both PHP 7.4 and 8.1 and could not recreate this. I tested with `adsensesetupV2` enabled and disabled. Mohit, would it be possible to recreate this with a screencast, and also include the plugins you have activated. Bit odd that I cannot recreate it at all.

**Comment by aaemnnosttv** [comment]
I just encountered this again during some unrelated testing. I can see that the account setup is marked as complete but the site setup still shows incomplete, even though the site status is `ready`. I'm going to move this forward as there is clearly something going on here but do some additional testing to see if I can see what is happening more concretely.

**Comment by aaemnnosttv** [comment]
I was able to reproduce this again, but not consistently. This short screencast shows the end result where you can see the module should be connected, but it is not.

https://user-images.githubusercontent.com/1621608/212766398-2d1a8cf2-cf08-44ca-8e81-1e0727d3386e.mp4

The problem seems to be within the setup flow but it's hard to see where.

When the settings start to get saved at the end, these are the requests I see (there are 4 in a row)

<details>
<summary>Request detail</summary>

1 – sets account ID, client ID as well as an `accountStatus` of `no-client`

```json
{
    "ownerID": 0,
    "accountID": "pub-12345",
    "autoAdsDisabled": [],
    "clientID": "ca-pub-12345",
    "accountStatus": "no-client",
    "siteStatus": "",
    "accountSetupComplete": false,
    "siteSetupComplete": false,
    "useSnippet": true,
    "webStoriesAdUnit": ""
}
```

2 – owner ID set from previous, `siteSetupComplete: true`, no `accountSetupComplete`

```json
{
    "ownerID": 1,
    "accountID": "pub-12345",
    "autoAdsDisabled": [],
    "clientID": "ca-pub-12345",
    "accountStatus": "no-client",
    "siteStatus": "",
    "accountSetupComplete": false,
    "siteSetupComplete": true,
    "useSnippet": true,
    "webStoriesAdUnit": ""
}
```

3 – duplicate of 2?

```json
{
    "ownerID": 1,
    "accountID": "pub-12345",
    "autoAdsDisabled": [],
    "clientID": "ca-pub-12345",
    "accountStatus": "no-client",
    "siteStatus": "",
    "accountSetupComplete": false,
    "siteSetupComplete": true,
    "useSnippet": true,
    "webStoriesAdUnit": ""
}
```

4 – `accountSetupComplete:true` and `siteSetupComplete:false` **both toggle**

```json
{
    "ownerID": 1,
    "accountID": "pub-12345",
    "autoAdsDisabled": [],
    "clientID": "ca-pub-12345",
    "accountStatus": "no-client",
    "siteStatus": "",
    "accountSetupComplete": true,
    "siteSetupComplete": false,
    "useSnippet": true,
    "webStoriesAdUnit": ""
}
```

</details>

### Restarting setup, via "Complete setup CTA"

<details>
<summary>Details</summary>

Requests accounts, sites, and clients as usual, and then saves settings automatically (note account and site statuses are `ready` but `siteSetupComplete` is still `false`:

```json
{
    "ownerID": 1,
    "accountID": "pub-12345",
    "autoAdsDisabled": [],
    "clientID": "ca-pub-12345",
    "accountStatus": "ready",
    "siteStatus": "ready",
    "accountSetupComplete": true,
    "siteSetupComplete": false,
    "useSnippet": true,
    "webStoriesAdUnit": ""
}
```

Clicking "Configure AdSense" triggers another settings request, but stays on the same step/state (note the module appears connected at this point, account and site setup complete flags are `true`)

```json
{
    "ownerID": 1,
    "accountID": "pub-12345",
    "autoAdsDisabled": [],
    "clientID": "ca-pub-12345",
    "accountStatus": "ready",
    "siteStatus": "ready",
    "accountSetupComplete": true,
    "siteSetupComplete": true,
    "useSnippet": true,
    "webStoriesAdUnit": ""
}
```

Clicking "Configure AdSense" for the second time triggers another request and invokes navigation to the dashboard (note `accountSetupComplete` now shows `false` again!)

```json
{
    "ownerID": 1,
    "accountID": "pub-12345",
    "autoAdsDisabled": [],
    "clientID": "ca-pub-12345",
    "accountStatus": "ready",
    "siteStatus": "ready",
    "accountSetupComplete": false,
    "siteSetupComplete": true,
    "useSnippet": true,
    "webStoriesAdUnit": ""
}
```

Another settings request is triggered immediately which sets `accountSetupComplete` back to `true`

```json
{
    "ownerID": 1,
    "accountID": "pub-12345",
    "autoAdsDisabled": [],
    "clientID": "ca-pub-12345",
    "accountStatus": "ready",
    "siteStatus": "ready",
    "accountSetupComplete": true,
    "siteSetupComplete": true,
    "useSnippet": true,
    "webStoriesAdUnit": ""
}
```

</details>

**Comment by techanvil** [comment]
I've been unable to reproduce this one.

I tried using the current release, 1.94.0, following the steps @mohitwp detailed [above](https://github.com/google/site-kit-wp/issues/5614#issuecomment-1357131786). I tried over ten times, resetting Site Kit between attempts. I then switched to release 1.90.0 which would have been in effect around the time Mohit was reproducing the issue and tried another ten times. Tried with devtools closed and open, with/without Slow 3G throttling, and with/without network caching enabled (although that shouldn't make a difference anyway, in theory).

As @aaemnnosttv mentioned it's a race condition, I wonder if it's contingent on the device being used. Having run the above tests on my regular Ubuntu workstation,I switched to an Intel MacBook Pro, and repeated the above with the same results.

Further details on the machine specs:
1. HP Probook 445 G8, with a Ryzen 7 5800U CPU and 32GB RAM running Ubuntu 22.04 and Chrome 110.0.5481.100 (64 bit).
2. MacBook Pro, 13-inch 2020, 2.3GHz i7, 16GB RAM, running macOS 13.1 and Chrome 110.0.5481.100 (x86_64).

@mohitwp / @aaemnnosttv, what sort of frequency were you seeing this issue occurring at, what spec machine are/were you running on, and can you still reproduce it?

**Comment by aaemnnosttv** [comment]
@techanvil hmm, that's interesting. I see this inconsistently, but in my testing above I seem to recall it was at least 1/10, maybe 1/5.

I'll see if I can find a way to improve that percentage, or ideally force it.

**Comment by jamesozzie** [comment]
I've been able to reproduce this on a subdomain site consistently both yesterday and today ([Asana task](https://app.asana.com/0/1185360131189553/1204289876526966/f)). In my case the parent domain is already approved for ads, so the subdomain site is already auto approved. 

https://user-images.githubusercontent.com/41326532/228605101-cec9cf28-859e-48f7-b342-d4f3b5c59587.mp4
 
  
  
@mohitwp @aaemnnosttv I'll keep this subdomain site ready for others who may wish to check this. One thing to note is that it also takes a couple of clicks on the "Configure Adense" button before the setup process proceeds. 

I also tested this on a main domain site (which is already approved). In this case the set up does complete successfully (although it also takes a couple of clicks on the "Configure AdSense" button during setup

https://user-images.githubusercontent.com/41326532/228605808-0ba0cdba-0145-42e0-b494-c26667ca69ea.mp4

**Comment by aaemnnosttv** [comment]
@bethanylang let's prioritize this as soon as we can fit it in. It's not severe enough to justify a P0 since the user isn't blocked (they can just run through it again which is quick) but I feel like it has been happening more recently.

**Comment by techanvil** [comment]
I've managed to pin down an apparent fix for this. This was a tricky and time consuming one to debug, due to its unpredictable and fairly low frequency occurrence (for me at any rate).

A couple of notes beyond what's been discussed above: I did find that I was able to reproduce this after a Site Kit reset, and also on my local dev environment. However I was only seeing the bug about 1 in 5 times when running through the setup.

It's worth mentioning I also noticed it behaved a bit differently on local compared to InstaWP, with some debugging tweaks seeming to fix it when running on the local environment but not when running on an InstaWP site. I think this is due to marginal timing modifications where the timing was different across the two environments.

Anyhow, I dug in and was able to pinpointed the documented fix which is working well for me on local and InstaWP.

**Comment by aaemnnosttv** [comment]
Thanks for digging into this @techanvil 👍 

The solution sounds reasonable, although it seems like this condition is already covered by the existing check for `canSubmitChanges` in the background submit effect: https://github.com/google/site-kit-wp/blob/fefb29197b90de7815b872a08f19311ea1e58a79/assets/js/modules/adsense/components/setup/v2/SetupMain.js#L187

Like other modules, the first thing checked is if changes are being submitted https://github.com/google/site-kit-wp/blob/276cae498615f7f4859be224d1451b0ab5703429/assets/js/modules/adsense/datastore/settings.js#L318

So adding a check for `isDoingSubmitChanges` to this effect seems like it would be redundant.

If this fixes it, I'd be interested to know why 😄 

The root issue seems to be that the background submit is reacting to changes that it shouldn't? The `complete*Setup` functions seem like they may easily be responsible for inadvertently submitting changes in the background every time it changes the status 🤔 
https://github.com/google/site-kit-wp/blob/fefb29197b90de7815b872a08f19311ea1e58a79/assets/js/modules/adsense/datastore/settings.js#L153-L166

**Comment by techanvil** [comment]
Hi @aaemnnosttv, many thanks for pointing that out. This really has not been the easiest one to debug, due to the sporadic nature and initially quite a slow path to reproducing it, although I did manage to speed that up a bit in the end.

So, it looks like I just got a bit lucky with the many runs I gave the `isDoingSubmitChanges` "fix"... Having given it an even more thorough run of testing I've discovered that it _doesn't_ work, so, as said, very good shout on that.

I did consider what you've pointed out regarding the `complete*Setup` functions, but, thinking I'd found a fix that didn't involve an additional flag I was happy to go with that. However, needs must and it looks like we should simply add another flag to prevent the background submission from happening while the form completion is in effect.

It's not the prettiest, but I've got a proposed fix in a PR, which I've linked in the IB. Please see what you think.

**Comment by aaemnnosttv** [comment]
Thanks @techanvil! Looking at the PR, I think we can use a less abstract name for the state as "completing setup" doesn't necessarily mean that background updates shouldn't be allowed. I'd suggest we rename that state to something like "suspend background submit" which is very explicit and easier to understand why it's there. This could be addressed during execution.

Regarding test coverage, it would be nice to cover this since I could see how it could potentially be reintroduced at some point due to the automatic nature of the background submit. Given that you have much more experience reproducing this now, do you think it's possible/reasonable to add coverage for this?

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, I've updated the IB to use the naming that you've suggested, SGTM.

With regard to the testing though - if you are talking about E2E test coverage, I don't really think it is possible to add coverage. It's not that I've found a way to reproduce the issue on demand, simply that I have found that the user flow can be short circuited a bit.

The initial bug reports suggested that it only occurs on the first setup of a new site. In fact I've found that simply deleting the AdSense settings and then running through the setup again is enough, and this can be reduced (on an InstaWP site) to a few quick steps - pressing a button on the DB web interface to trigger the SQL deletion, and then navigating to a bookmark to the AdSense setup page to press the CTA.

However, the bug still only occurs about 1 in 5 times in my testing, which makes it unsuitable for an E2E test. _Maybe_ there's some way of setting up the right conditions e.g. by delaying a particular response but it's not something I've come across during my somewhat lengthy investigation.

As for component/unit testing, it appears we currently don't have _any_ for the whole set of V2 AdSense components, so I don't think it's worthwhile introducing a complicated test purely for this edge case...

**Comment by aaemnnosttv** [comment]
Sounds good, thanks. I was mostly thinking about something at the component level where we could force the race condition, by controlling when a request finishes, etc but it isn't worth sinking a ton of time into.

IB ✅

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Tested on dev environment.
- Verified set up through CTA and from settings.
- AdSense set up is getting complete in both scenarios.
- No unexpected behavior and console error found.

<details>

https://github.com/google/site-kit-wp/assets/94359491/c8f87647-835f-4aba-853e-a780d90affe5

</details>
